File: runtime/python/units/wasi/path_readlink.py
# requires: memory/read_string, memory/init, memory/i32_store, wasi/resolve_path, wasi/errno_fs
def wasi_path_readlink(self, fd, path_ptr, path_len, buf_ptr, buf_len, bufused_ptr):
    rel = self.memory.read_string(path_ptr, path_len).decode("utf-8", "surrogateescape")
    host, err = self.resolve_path(fd, rel, False)
    if err is not None:
        return err
    try:
        target = os.readlink(host)
    except OSError as e:
        return self.fs_errno(e)
    data = target.encode("utf-8", "surrogateescape")
    # The content is truncated to buf_len (no NUL terminator); bufused reports
    # how many bytes were written.
    n = min(len(data), buf_len)
    self.memory.init(buf_ptr, data, 0, n)
    self.memory.i32_store(bufused_ptr, n)
    return self.ERRNO_SUCCESS

File: runtime/ruby/units/wasi/path_readlink.rb
# requires: memory/read_string, memory/init, memory/i32_store, wasi/resolve_path, wasi/errno_fs
def wasi_path_readlink(dirfd, path_ptr, path_len, buf_ptr, buf_len, bufused_ptr)
  rel = @memory.read_string(path_ptr, path_len)
  # readlink operates on the link itself: resolve the parent but not the
  # final component.
  host_path, err = resolve_path(dirfd, rel, follow_last: false)
  return err if err
  bytes = File.readlink(host_path).b # raises EINVAL when not a symlink
  n = [bytes.bytesize, buf_len].min
  @memory.init(buf_ptr, bytes, 0, n)
  @memory.i32_store(bufused_ptr, n)
  ERRNO_SUCCESS
rescue SystemCallError => e
  fs_errno(e)
end
